framework/graph/GraphBuilder.kt

```diff
@@ -399,26 +465,70 @@ class GraphBuilder(
-                    val relationTypeLabel = roleType.label.scope().get()
-                    val roleLabel = roleType.label.name()
-                    val relationTypeVertex = graphBuilder.allTypeVertices[relationTypeLabel] as? Vertex.Type.Relation
-                    if (relationTypeVertex != null) {
-                        graphBuilder.addEdge(Edge.Plays(relationTypeVertex, typeVertex, roleLabel))
-                    } else {
-                        graphBuilder.addEdgeCandidate(EdgeCandidate.Plays(relationTypeLabel, typeVertex, roleLabel))
-                    }
+                graphBuilder.playsScopedEdgeCandidatesLock.writeLock().withLock {
+                    remoteThingType?.plays
+                        ?.forEach { roleType ->
+                            val relationTypeLabel = roleType.label.scope().get()
+                            val roleLabel = roleType.label.name()
+                            val roleRelationPair = Pair(relationTypeLabel, roleLabel).toString()
+                            if (!graphBuilder.playsRoleScopedEdgeCandidates.containsKey(roleRelationPair)) {
+                                val scopedEdgeCand = ScopedEdgeCandidate.Plays(relationTypeLabel, typeVertex, roleType, getSupertypeList(remoteThingType!!))
+                                graphBuilder.playsRoleScopedEdgeCandidates[roleRelationPair] = listOf(scopedEdgeCand)
+                            }
+
+                            var supertypeFound = false
+                            val ownsScopedEdgeCandidates = graphBuilder.playsRoleScopedEdgeCandidates[roleRelationPair]!!
+                            ownsScopedEdgeCandidates.forEach {
+                                if (it.hasSupertype(thingType)) {
+                                    it.rescope(typeVertex)
+                                }
+                                else if (remoteThingType!!.supertypes.toList().toSet().contains(it.target.concept.asThingType())) {
+                                    supertypeFound = true
+                                }
+                            }
+                            if (!supertypeFound) {
+                                val scopedEdgeCand = ScopedEdgeCandidate.Plays(relationTypeLabel, typeVertex, roleType, getSupertypeList(remoteThingType!!))
+                                graphBuilder.playsRoleScopedEdgeCandidates[roleRelationPair] =
+                                    graphBuilder.playsRoleScopedEdgeCandidates[roleRelationPair]!! + listOf(scopedEdgeCand)
+                            }
+                        }
                 }
             }
+
+            private fun getSupertypeList(remoteThingType: Remote): List<com.vaticle.typedb.client.api.concept.type.ThingType> {
+                return remoteThingType.supertypes
+                    .filter { it.label.name() !in listOf("thing", "entity", "relation", "attribute", remoteThingType.label.name()) }
+                    .toList()
+            }
```

Comment:
The root types can't own attributes or play roles. We also don't want ourselves in our list of supertypes for this purpose.